### ISSUE TITLE
Some issues on Windows

### ISSUE THREAD

**Issue by divinon** [polymonster/pmtech#36]
Hello, I had some fun and problems on Windows I'd like to share, maybe it is of some use.

If build.py is called without args then the variabile _info is not initialized.
Solved by calling `_info = build_info()` in get_platform_info.

In build.py, for Windows it could be helpful to set 'python_exec' to 'py -3' in case you have both pythons.

In someway or another I had an empty info.json file in examples\bin\win32\data\pmfx\hlsl\compute, this caused the json parser to crash.
Solved by adding a check on the file size in check_dependencies of build_pmfx.py: 
`    if os.path.exists(info_filename) and os.path.getsize(info_filename) > 0:
`

The compute shader was a bit tricky: calling build without args will default to HLSL 4.0, this causes the shader greyscale to be skipped since it has "supported_platforms 5.0". Here a warning for the skipped shader could be useful, otherwise all seems good but the compute example will crash because there is no technique in the shader.
By setting -shader_version to 5_0 I was able to compile it, the sample was running but I had a black screen with the error "d3d device error: The parameter is incorrect." and:

D3D11 ERROR: ID3D11Device::CreateTexture2D: The format (0x57, B8G8R8A8_UNORM) cannot be bound as an UnorderedAccessView, or cast to a format that could be bound as an UnorderedAccessView.  Therefore this format does not support D3D11_BIND_UNORDERED_ACCESS. [ STATE_CREATION ERROR #92: CREATETEXTURE2D_UNSUPPORTEDFORMAT]
Exception thrown at 0x00007FFCC9F792FC in basic_compute_d.exe: Microsoft C++ exception: _com_error at memory location 0x000000338F8EDA30.
D3D11 ERROR: ID3D11Device::CreateTexture2D: Returning E_INVALIDARG, meaning invalid parameters were passed. [ STATE_CREATION ERROR #104: CREATETEXTURE2D_INVALIDARG_RETURN]
D3D11 ERROR: ID3D11Device::CreateShaderResourceView: A ShaderResourceView cannot be created from a NULL Resource. [ STATE_CREATION ERROR #129: CREATESHADERRESOURCEVIEW_INVALIDRESOURCE]
Exception thrown at 0x00007FFCC9F792FC in basic_compute_d.exe: Microsoft C++ exception: _com_error at memory location 0x000000338F8EDCA8.
D3D11 ERROR: ID3D11Device::CreateShaderResourceView: Returning E_INVALIDARG, meaning invalid parameters were passed. [ STATE_CREATION ERROR #131: CREATESHADERRESOURCEVIEW_INVALIDARG_RETURN]
D3D11 ERROR: ID3D11Device::CreateUnorderedAccessView: A UnorderedAccessView cannot be created from a NULL Resource. [ STATE_CREATION ERROR #2097341: CREATEUNORDEREDACCESSVIEW_INVALIDRESOURCE]
Exception thrown at 0x00007FFCC9F792FC in basic_compute_d.exe: Microsoft C++ exception: _com_error at memory location 0x000000338F8EDC00.
D3D11 ERROR: ID3D11Device::CreateUnorderedAccessView: Returning E_INVALIDARG, meaning invalid parameters were passed. [ STATE_CREATION ERROR #2097351: CREATEUNORDEREDACCESSVIEW_INVALIDARG_RETURN]

From here: https://docs.microsoft.com/en-us/previous-versions//ff471325(v=vs.85)#structured-uav-and-srv
I get that only DXGI_FORMAT_R8G8B8A8_UNORM is supported. Changing it in dds_pixel_format_to_texture_format solved this.

**Comment by polymonster** [comment]
Thanks for letting me know about this, I have been running mostly on Mac so windows sometimes falls a little behind and until recently not many people have been using this code base so it has just been me keeping it up to date.

For python on windows if using cmd prompt I am always using "py -3 build.py -all".. I am not sure what you mean by set 'python_exec' to 'py -3'

I did have a think about shader compilation recently and I think the best solution for hlsl would be instead of excluding shaders from compilation, the pmfx config specifies which shader version to target. I want to keep most the shaders as 4_0 as I have a device which doesn't support sm 5_0. So in the compute shader case 5_0 would be compiled and then at run-time I would prevent the sample from running with an assert or something nicer than just crash!

I need to take a look on windows at the D3D11 validation issue, but will fix the errors.

I will get back to you once I have fixed these.

**Comment by polymonster** [comment]
@divinon I have fixed the following issues:

1. running build.py without args _info is now initialised in main
2. when specifying supported platforms in pmfx you will get compiled binary for that shader model, so you do not have to run -shader_version unless you want to force all shaders to that shader model.

I was unable to reproduce the issue with d3d11 errors, when I run the sample it works as expected with no errors, so I think it must be to do with the d3d driver, what GPU do you have? I am running an nvidia gtx 1060... I will try running my work computer tomorrow to see if there is any difference.

**Comment by divinon** [comment]
Thank you. Now compute works, and the shader is compiled with no issue.
I was or on a GTX 860M or on an Intel HD 4600, not sure anymore, probably the latter. Now I've printed the dxgi adapter description and I'm sure it works on both the cards.

About 'python_exec', inside build.py the command passed to subprocess to, for example, compile the shaders is: `build_steps.append(python_exec + " " + shader_script + " " + shader_options)`  
in Windows that 'python_exec' is an empty string, so the default application for .py files is used and it could be python 2.7.  It is here:
https://github.com/polymonster/pmtech/blob/01d693afd8a18cb6174b452b844a39b6f6c9d6eb/tools/build.py#L149

And the file size check was added here (the json parser error was not friendly to understand):
https://github.com/polymonster/pmtech/blob/01d693afd8a18cb6174b452b844a39b6f6c9d6eb/tools/build_scripts/build_pmfx.py#L229
